### ISSUE TITLE
Setting OMP_NUM_THREADS does not affect org.nd4j.nativeblas.NativeOps - Number of threads used for linear algebra

### ISSUE THREAD

**Issue by rasoolims** [deeplearning4j/deeplearning4j#1701]
When I change export OMP_NUM_THREADS=10, I still see that num of threads is one!

> 03:43:34.520 [main] DEBUG org.nd4j.nativeblas.NativeOps - Number of threads used for linear algebra 1
> 03:43:34.530 [main] DEBUG org.nd4j.nativeblas.NativeOps - Number of threads used for linear algebra 1
> 03:43:34.610 [main] DEBUG org.reflections.Reflections - going to scan these urls:
> jar:file:/proj/mlnlp/rasooli/code/neural_blas_test/target/YaraParser-0.1-bin.jar!/
> 03:43:35.129 [main] DEBUG org.reflections.Reflections - could not scan file org/nd4j/nativeblas/linux-x86_64/libgomp.so.1 in url jar:file:/proj/mlnlp/rasooli/code/neural_blas_test/target/YaraParser-0.1-bin.jar!/ with scanner TypeAnnotationsScanner
> 03:43:35.131 [main] DEBUG org.reflections.Reflections - could not scan file org/nd4j/nativeblas/linux-x86_64/libgomp.so.1 in url jar:file:/proj/mlnlp/rasooli/code/neural_blas_test/target/YaraParser-0.1-bin.jar!/ with scanner SubTypesScanner
> 03:43:35.132 [main] DEBUG org.reflections.Reflections - could not scan file org/nd4j/nativeblas/linux-x86_64/libgfortran.so.3 in url jar:file:/proj/mlnlp/rasooli/code/neural_blas_test/target/YaraParser-0.1-bin.jar!/ with scanner TypeAnnotationsScanner
> 03:43:35.132 [main] DEBUG org.reflections.Reflections - could not scan file org/nd4j/nativeblas/linux-x86_64/libgfortran.so.3 in url jar:file:/proj/mlnlp/rasooli/code/neural_blas_test/target/YaraParser-0.1-bin.jar!/ with scanner SubTypesScanner
> 03:43:35.132 [main] DEBUG org.reflections.Reflections - could not scan file org/nd4j/nativeblas/linux-x86_64/libopenblas.so.0 in url jar:file:/proj/mlnlp/rasooli/code/neural_blas_test/target/YaraParser-0.1-bin.jar!/ with scanner TypeAnnotationsScanner
> 03:43:35.132 [main] DEBUG org.reflections.Reflections - could not scan file org/nd4j/nativeblas/linux-x86_64/libopenblas.so.0 in url jar:file:/proj/mlnlp/rasooli/code/neural_blas_test/target/YaraParser-0.1-bin.jar!/ with scanner SubTypesScanner
> 03:43:35.133 [main] DEBUG org.reflections.Reflections - could not scan file org/nd4j/nativeblas/linux-x86_64/libnd4j.so in url jar:file:/proj/mlnlp/rasooli/code/neural_blas_test/target/YaraParser-0.1-bin.jar!/ with scanner TypeAnnotationsScanner
> 03:43:35.133 [main] DEBUG org.reflections.Reflections - could not scan file org/nd4j/nativeblas/linux-x86_64/libnd4j.so in url jar:file:/proj/mlnlp/rasooli/code/neural_blas_test/target/YaraParser-0.1-bin.jar!/ with scanner SubTypesScanner
> 03:43:35.133 [main] DEBUG org.reflections.Reflections - could not scan file org/nd4j/nativeblas/linux-x86_64/libjnind4j.so in url jar:file:/proj/mlnlp/rasooli/code/neural_blas_test/target/YaraParser-0.1-bin.jar!/ with scanner TypeAnnotationsScanner
> 03:43:35.134 [main] DEBUG org.reflections.Reflections - could not scan file org/nd4j/nativeblas/linux-x86_64/libjnind4j.so in url jar:file:/proj/mlnlp/rasooli/code/neural_blas_test/target/YaraParser-0.1-bin.jar!/ with scanner SubTypesScanner

I also extracted the jar file and went to org/nd4j/nativeblas/linux-x86_64:

```bash
ls -lah 
```

> drwx------ 2 rasooli phd 4.0K Jun 16 03:23 .
drwx------ 3 rasooli phd 4.0K Jun 16 03:23 ..
-rw------- 1 rasooli phd 965K Jun 16 03:23 libgfortran.so.3
-rw------- 1 rasooli phd  84K Jun 16 03:23 libgomp.so.1
-rw------- 1 rasooli phd 173K Jun 16 03:23 libjnind4j.so
-rw------- 1 rasooli phd 1.4M Jun 16 03:23 libnd4j.so
-rw------- 1 rasooli phd  33M Jun 16 03:23 libopenblas.so.0

How should I fix this?

**Comment by treo** [comment]
The bug here is in the message, not in the actual CPU usage. If you take a look at your cpu usage, you will see that it does use 10 threads.

**Comment by rasoolims** [comment]
as a follow-up
```java
System.out.println("OMP_NUM_THREADS "+  System.getenv().get("OMP_NUM_THREADS"));
```

> 48

I am trying to train a deep MLP with 32 embedding layers (12 have dimension of 64 and others 32), two hidden layers of size 200 and a softmax of size 22 possible outputs (with batchsize of 10k). When I start to train with 48 threads by setting OMP_NUM_THREADS=48 (my machine has 24 cores), based on what ScoreIterationListener reports, each iteration takes 30 seconds (my data has about 2M instances). Am I doing sth wrong that this takes this much?

Here is the gist for that
https://gist.github.com/rasoolims/5268edebcd2882e8d07687967bb3654b

Also, here is the cpu usage over time:
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND  
54879 rasooli   30  10 26.566g 3.044g  17124 S **187.1**  3.2   0:23.80 java                                                                                                                                                                                                     
54879 rasooli   30  10 26.566g 3.446g  17140 S **328.2**  3.7   0:33.69 java                                                                                                                                                                                                     
54879 rasooli   30  10 26.566g 3.453g  17140 S **106.1**  3.7   0:36.89 java                                                                                                                                                                                                     
54879 rasooli   30  10 26.566g 3.463g  17140 S **108.5**  3.7   0:40.16 java                                                                                                                                                                                                     
54879 rasooli   30  10 26.566g 4.973g  17140 S **101.5**  5.3   0:43.22 java                                                                                                                                                                                                     
54879 rasooli   30  10 26.566g 4.973g  17140 S **100.2**  5.3   0:46.24 java                                                                                                                                                                                                     
54879 rasooli   30  10 26.566g 4.973g  17140 S  **99.9**  5.3   0:49.25 java                                                                                                                                                                                                     
54879 rasooli   30  10 26.663g 5.076g  17280 S **117.5**  5.4   0:52.79 java                                                                                                                                                                                                     
54879 rasooli   30  10 26.777g 5.186g  17280 S **104.5**  5.5   0:55.94 java                                                                                                                                                                                                     
54879 rasooli   30  10 26.720g 5.129g  17292 S **191.4**  5.4   1:01.71 java                                                                                                                                                                                                     
54879 rasooli   30  10 27.071g 5.163g  17344 S **134.5**  5.5   1:05.76 java                                                                                                                                                                                                     
54879 rasooli   30  10 27.224g 5.247g  17400 S **111.9**  5.6   1:09.13 java                                                                                                                                                                                                     
54879 rasooli   30  10 27.383g 5.451g  17472 S **146.5**  5.8   1:13.54 java                                                                                                                                                                                                     
54879 rasooli   30  10 27.376g 5.485g  17480 S **129.3**  5.8   1:17.44 java                                                                                                                                                                                                     
54879 rasooli   30  10 27.438g 5.498g  17480 S **108.8**  5.8   1:20.72 java                                                                                                                                                                                                     
54879 rasooli   30  10 27.438g 5.578g  17480 S **207.6**  5.9   1:26.97 java                                                                                                                                                                                                     
54879 rasooli   30  10 27.438g 5.608g  17496 S **113.2**  5.9   1:30.38 java                                                                                                                                                                                                     
54879 rasooli   30  10 27.501g 7.128g  17496 S **157.2**  7.6   1:35.12 java                                                                                                                                                                                                     
54879 rasooli   30  10 27.563g 7.144g  17496 S **101.5**  7.6   1:38.18 java                                                                                                                                                                                                     
54879 rasooli   30  10 27.626g 7.224g  17496 S **100.2**  7.7   1:41.20 java                                                                                                                                                                                                     
54879 rasooli   30  10 27.626g 7.224g  17496 S **110.5**  7.7   1:44.53 java                                                                                                                                                                                                     
54879 rasooli   30  10 27.678g 7.428g  17512 S **116.1**  7.9   1:48.03 java                                                                                                                                                                                                     
54879 rasooli   30  10 27.678g 8.688g  17512 S **108.1**  9.2   1:51.29 java                                                                                                                                                                                                     
54879 rasooli   30  10 27.731g 8.997g  17512 S **120.1**  9.5   1:54.91 java                                                                                                                                                                                                     
54879 rasooli   30  10 27.731g 8.997g  17512 S **109.1**  9.5   1:58.20 java                                                                                                                                                                                                     
54879 rasooli   30  10 27.793g 9.018g  17512 S **109.3**  9.6   2:01.49 java                                                                                                                                                                                                     
54879 rasooli   30  10 27.793g 9.050g  17512 S **111.1**  9.6   2:04.84 java                                                                                                                                                                                                     
54879 rasooli   30  10 27.793g 9.130g  17512 S **229.5**  9.7   2:11.76 java

**Comment by treo** [comment]
1. Use only as much threads as you have actual physical cores (i.e. not as many as you have hyper threads, and esp. not more than that).
2. If you are looking for the very best performance, you should be using MKL, and compile from source on a pretty recent Linux (one that has glibc >= 2.22).

**Comment by raver119** [comment]
Could you please provide hot spots screenshot, using any java profiler?

**Comment by rasoolims** [comment]
I changed OMP_NUM_THREADS to 10 to make sure that it is less that the number of my cores. I profiled it remotely with VisualVM

![1](https://cloud.githubusercontent.com/assets/1611317/16155247/780bc352-347c-11e6-871d-42d0182d9164.jpg)

![2](https://cloud.githubusercontent.com/assets/1611317/16155253/7d643474-347c-11e6-9d2d-34c0ef726be5.png)

Maybe my blas does not function properly!  
``` 
ldconfig -p | grep blas
```

> libgslcblas.so.0 (libc6,x86-64) => /usr/lib/libgslcblas.so.0
> libgslcblas.so (libc6,x86-64) => /usr/lib/libgslcblas.so
> libblas.so.3 (libc6,x86-64) => /usr/lib/libblas.so.3
> libblas.so (libc6,x86-64) => /usr/lib/libblas.so


A locally installed OpenBLAS is also added to my path

**Comment by treo** [comment]
It would be great if you opened up that `gradientAndScore` down to the methods that take up most of the time.

**Comment by raver119** [comment]
Uhhh... That definitely looks bad. Please provide screenshot with expanded tree, as @treo asks.

**Comment by rasoolims** [comment]
Sorry, the whole thing did not fit into screen but I tried to cover the main consuming parts

![3](https://cloud.githubusercontent.com/assets/1611317/16156909/4a46bba4-3484-11e6-9b80-50f018cbea81.png)

**Comment by raver119** [comment]
So, you have ENFORCE_NUMERICAL_STABILITY somewhere?

**Comment by rasoolims** [comment]
Before starting anything I had:
```java
Nd4j.ENFORCE_NUMERICAL_STABILITY = true;
```

**Comment by raver119** [comment]
Disable that then. That thing works in single thread. And we definitely want it to become multithreaded, since it exists.

**Comment by rasoolims** [comment]
Great, it is now between 300% to 900% percent cpu usage by choosing OMP_NUM_THREADS=**24**(=num of cores); though it is still less than 2000%. When I choose e.g. OMP_NUM_THREADS=**10** cpu usage decreases proportionally

**Comment by raver119** [comment]
Great.

Send new hot spots screenshot then.

**Comment by rasoolims** [comment]
![5](https://cloud.githubusercontent.com/assets/1611317/16158109/babc90e8-3489-11e6-9def-501bb2533ab0.png)

![6](https://cloud.githubusercontent.com/assets/1611317/16158131/d5fed73a-3489-11e6-96a0-ff5552c14b46.png)

**Comment by raver119** [comment]
Well, this looks +- expectable now. EmbeddingLayer has single-threaded code inside, and your network is built as bunch of those...

**Comment by rasoolims** [comment]
Is there any way to force EmbeddingLayer to be multi-threaded as well?

**Comment by raver119** [comment]
We're discussing that possibility.

**Comment by rasoolims** [comment]
BTW, I am also pretty sure that the prediction phase is still single-threaded

```java
 while (devIter.hasNext()) {
    MultiDataSet t = devIter.next();
    INDArray[] features = t.getFeatures();
    INDArray[] predicted = net.output(false,features);
}
```

**Comment by raver119** [comment]
Prediction phase operates on 1-Dimensional input i guess, not too much room for mulithreading there, you know :)

But there's still SIMD possible. to get support for SIMD stuff, you'll have to use either Intel MKL, or modern Linux distribution, with glibc 2.22+ and build libnd4j/nd4j from sources manually.

**Comment by raver119** [comment]
Actually, output() method accepts batched inputs as well, so if you have multiple features available - pass them in as batch, and you'll get outputs all at once. With internal multithreading etc.

**Comment by rasoolims** [comment]
You are right about batch inputs but for certain applications (such as my work for dependency parsing) it is quite impossible to have batch inputs in large sizes. So if DL4J can have such an ability to benefit from multi-core machines that would be great.

In almost all natural language processing tasks, there are bunch of embedding layers so it is must for NLP to have multi-threaded embedding layer. I will wait for the next release.
